geometry/eolearn/geometry/transformations.py

```diff
@@ -134,15 +145,41 @@ def _get_vector_data(self, eopatch):
         feature_type, feature_name = next(self.vector_input(eopatch))
         return eopatch[feature_type][feature_name]
 
-    def _get_rasterization_shapes(self, eopatch, group_classes=False):
+    def _sort_timestamps(self, eopatch):
+        """ indexes all timestamps found in the vector input data according to the eopatch timestamps. Any 
+        timestamps present in the vector data but missing from the EOPatch are ignored. The exception to this
+        is if no timestamps are found in the EOPatch, they are added from the vector data. 
+        """
+        vectors = self._get_vector_data(eopatch)
+        vector_timestamps = set(vectors['TIMESTAMP'])
+
+        if len(eopatch.timestamp )==0:
+            eopatch.timestamp = list(vector_timestamps)
+            warnings.warn('The EOPatch has no timestamp data. Timestamps from the vector data will be added '
+                          'to the EOPatch', RuntimeWarning)
+
+        if any([t_vec not in eopatch.timestamp for t_vec in vector_timestamps]):
+            warnings.warn('Some timestamps in the vector data are not found in the eopatch timestamps and will be '
+                          'ignored', RuntimeWarning)
+
+        timestamp_flags = [(idx, time, 1) if time in vector_timestamps else (idx, time, 0) for idx, time in enumerate(eopatch.timestamp)]
+
+        return timestamp_flags
+
+    def _get_rasterization_shapes(self, eopatch, group_classes=False, timestamp=None):
         """ Returns a generator of pairs of geometrical shapes and values. In rasterization process each shape will be
         rasterized to it's corresponding value.
         If there are no such geometries it will return `None`
         :param group_classes: If true, function returns a zip that iterates through cascaded unions of polygons of the
         same class, otherwise zip iterates through all polygons regardless of their class.
         :type group_classes: boolean
+        :param timestamp: If supplied, only shapes associated with the given timestamp will be returned.
+        :type timestamp: datetime.datetime
         """
         vector_data = self._get_vector_data(eopatch)
+        
+        if timestamp:
+            vector_data = vector_data[vector_data['TIMESTAMP']==timestamp]
```

Comment:
This can be quite inefficient if there are a lot of timestamps. This would require a bit more refactoring but I suggest something like this:

1. remove timestamps that are not in `eopatch.timestamp` (e.g. with `isin` method),
2. use `_get_rasterization_shapes` on the remaining dataframe (without splitting it by timestamp yet),
3. before rasterization split the final dataframe by timestamps. I think `pandas` `groupby` operation should be more efficient in splitting by timestamp than the current implementation.

Reply:
I followed step 1 and 2. For step 3 I used `itertools` `groupby` instead of `pandas` since the `_get_rasterization_shapes` returns a zip object.